File: ios/Classes/FlutternativedataPlugin.swift
import Flutter
import UIKit

public class FlutternativedataPlugin: NSObject, FlutterPlugin {
    
    enum CallMethod {
        case getPlatformVersion
        case getBatteryLevel
        case getDeviceInfo
        case getMemoryInfo
        case getPackageInfo
        
        public var rawValue: String {
            switch self {
            case .getPlatformVersion:
                "getPlatformVersion"
            case .getBatteryLevel:
                "getBatteryLevel"
            case .getDeviceInfo:
                "getDeviceInfo"
            case .getMemoryInfo:
                "getMemoryInfo"
            case .getPackageInfo:
                "getPackageInfo"
            }
        }
    }
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "flutternativedata", binaryMessenger: registrar.messenger())
        let instance = FlutternativedataPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    /// Called if this plugin has been registered to receive `FlutterMethodCall`s.
    /// - Parameters:
    ///   - call: The method call command object.
    ///   - result: A callback for submitting the result of the call.
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let callMethod = CallMethod.self
        switch call.method {
        case callMethod.getPlatformVersion.rawValue:
            result("iOS " + UIDevice.current.systemVersion)
        case callMethod.getBatteryLevel.rawValue:
            UIDevice.current.isBatteryMonitoringEnabled = true
            let batteryLevel = UIDevice.current.batteryLevel
            let batteryLevelPercentage = Int(batteryLevel * 100)
            result(batteryLevelPercentage)
        case callMethod.getDeviceInfo.rawValue:
            result(getDeviceInfo())
        case callMethod.getMemoryInfo.rawValue:
            result(getMemoryInfo())
        case callMethod.getPackageInfo.rawValue:
            result(getPackageInfo())
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
  private func getDeviceInfo() -> [String: Any] {
    var deviceInfo = [String: Any]()
    deviceInfo["deviceModel"] = UIDevice.current.model
    deviceInfo["deviceName"] = UIDevice.current.name
    deviceInfo["systemVersion"] = UIDevice.current.systemVersion
    deviceInfo["localizedModel"] = UIDevice.current.localizedModel
    deviceInfo["identifierForVendor"] = UIDevice.current.identifierForVendor?.uuidString
    deviceInfo["deviceType"] = UIDevice.current.userInterfaceIdiom == .phone ? "Phone" : "Tablet"
    deviceInfo["deviceLanguage"] = Locale.preferredLanguages.first ?? "Unknown"
    deviceInfo["deviceCountry"] = Locale.current.regionCode ?? "Unknown"
    deviceInfo["deviceTimeZone"] = TimeZone.current.identifier
    deviceInfo["deviceManufacturer"] = "Apple" // Added
    deviceInfo["deviceBrand"] = "Apple" // Added
    deviceInfo["deviceProduct"] = "iOS Device" // Added
    deviceInfo["deviceHardware"] = "iOS Hardware" // Added
    deviceInfo["deviceOSVersion"] = UIDevice.current.systemVersion // Updated
    deviceInfo["deviceSDKVersion"] = UIDevice.current.systemVersion // Updated
    deviceInfo["deviceID"] = UIDevice.current.identifierForVendor?.uuidString ?? "Unknown" // Updated
    return deviceInfo
}


    
    private func getPackageInfo() -> [String: Any]? {
        guard let infoDictionary = Bundle.main.infoDictionary else {
            return nil
        }
        
        var packageInfo = [String: Any]()
        packageInfo["versionName"] = infoDictionary["CFBundleShortVersionString"] as? String ?? "N/A"
        packageInfo["versionCode"] = infoDictionary["CFBundleVersion"] as? String ?? "N/A"
        packageInfo["packageName"] = Bundle.main.bundleIdentifier ?? "N/A"
        
        return packageInfo
    }
    
    private func getMemoryInfo() -> [String: Any] {
        var memoryInfo = [String: Any]()
        
        // Get total memory
        // Get total memory
        let totalMemory = ProcessInfo.processInfo.physicalMemory
        memoryInfo["totalMemory"] = totalMemory
        
        // Get available memory
        var pageSize: vm_size_t = 0
        host_page_size(mach_host_self(), &pageSize)
        var vmStat: vm_statistics_data_t = vm_statistics_data_t()
        var count = mach_msg_type_number_t(MemoryLayout<vm_statistics_data_t>.size / MemoryLayout<integer_t>.size)
        let status = withUnsafeMutablePointer(to: &vmStat) {
            $0.withMemoryRebound(to: integer_t.self, capacity: Int(count)) {
                host_statistics(mach_host_self(), HOST_VM_INFO, $0, &count)
            }
        }
        if status == KERN_SUCCESS {
            let freeMemory = UInt64(vmStat.free_count) * UInt64(pageSize)
            let usedMemory = UInt64(vmStat.active_count + vmStat.inactive_count + vmStat.wire_count) * UInt64(pageSize)
            memoryInfo["availableMemory"] = freeMemory
            memoryInfo["usedMemory"] = usedMemory
        } else {
            memoryInfo["availableMemory"] = "N/A"
            memoryInfo["usedMemory"] = "N/A"
        }
        return memoryInfo
    }
}

File: android/src/main/kotlin/com/example/flutternativedata/FlutternativedataPlugin.kt
package com.example.flutternativedata
import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.os.BatteryManager
import android.os.Build
import android.app.ActivityManager
import java.util.Locale
import java.util.TimeZone
import android.provider.Settings
import android.telephony.TelephonyManager
import androidx.core.app.ActivityCompat
import android.net.wifi.WifiManager
import java.net.NetworkInterface
import java.util.Collections

import androidx.annotation.NonNull

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** FlutternativedataPlugin */
class FlutternativedataPlugin: FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android

   // Declare context variable
    private lateinit var context: Context
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel
  private lateinit var batteryManager: BatteryManager
  private lateinit var packageManager: PackageManager
  private lateinit var activityManager: ActivityManager


  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutternativedata")
    channel.setMethodCallHandler(this)

   // Initialize context and other variables
    context = flutterPluginBinding.applicationContext
    packageManager = context.packageManager
    batteryManager = flutterPluginBinding.applicationContext.getSystemService(Context.BATTERY_SERVICE) as BatteryManager
    activityManager = flutterPluginBinding.applicationContext.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    when (call.method) {
      "getBatteryLevel" -> {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
          result.success(batteryManager.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY));
        } else {
          result.error("WRONG_VERSION", "android version not supported", "");
        }
      }
      "getPlatformVersion" -> {
        // Implement the logic to retrieve the platform version
        val version = "Android ${android.os.Build.VERSION.RELEASE}"
        result.success(version)
      }
      "getDeviceInfo" -> {
        val deviceInfo = getDeviceInfo()
                result.success(deviceInfo)
            }
     "getPackageInfo" -> {
        val packageInfo = getPackageInfo()
                result.success(packageInfo)
            }
      "getMemoryInfo" -> {
          val memoryInfo = getMemoryInfo()
                result.success(memoryInfo)
            }
      else -> {
        result.notImplemented()
      }
    }
  }

 private fun getPackageInfo(): Map<String, Any>? {
    val infoMap = mutableMapOf<String, Any>()
     val packageManager = context.packageManager
     val packageInfo = packageManager.getPackageInfo(context.packageName, 0)
    val applicationInfo = packageManager.getApplicationInfo(context.packageName, 0)
    val appName = packageManager.getApplicationLabel(applicationInfo).toString()

    try {
        val packageInfo = packageManager.getPackageInfo(context.packageName, 0)
        infoMap["versionName"] = packageInfo.versionName
        infoMap["versionCode"] = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            packageInfo.longVersionCode
        } else {
            packageInfo.versionCode
        }
        infoMap["packageName"] = context.packageName
        infoMap["appName"] =  appName 
    } catch (e: PackageManager.NameNotFoundException) {
        // Handle exception if package information is not found
        return null
    }
    return infoMap
}

fun getSecondIMEI(context: Context): String? {
    val telephonyManager = context.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
    return if (ActivityCompat.checkSelfPermission(context, Manifest.permission.READ_PHONE_STATE) == PackageManager.PERMISSION_GRANTED) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            telephonyManager.getImei(1) // 1 for the second SIM slot
        } else {
            null
        }
    } else {
        null
    }
}

fun getMacAddress(context: Context): String? {
    val wifiManager = context.applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
    return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
        try {
            val all = Collections.list(NetworkInterface.getNetworkInterfaces())
            for (nif in all) {
                if (nif.name.equals("wlan0", ignoreCase = true)) {
                    val macBytes = nif.hardwareAddress ?: return null
                    val macAddress = StringBuilder()
                    for (b in macBytes) {
                        macAddress.append(String.format("%02X:", b))
                    }
                    if (macAddress.isNotEmpty()) {
                        macAddress.deleteCharAt(macAddress.length - 1)
                    }
                    return macAddress.toString()
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        null
    } else {
        val wifiInfo = wifiManager.connectionInfo
        wifiInfo.macAddress
    }
}


fun getDeviceSerialNumber(): String {
    return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
        Build.getSerial()
    } else {
        Build.SERIAL
    }
}


private fun getDeviceInfo(): Map<String, Any> {
    val deviceInfoMap = mutableMapOf<String, Any>()
    deviceInfoMap["deviceModel"] = Build.MODEL
    deviceInfoMap["deviceManufacturer"] = Build.MANUFACTURER
    deviceInfoMap["deviceBrand"] = Build.BRAND
    deviceInfoMap["deviceName"] = Build.DEVICE
    deviceInfoMap["deviceProduct"] = Build.PRODUCT
    deviceInfoMap["deviceHardware"] = Build.HARDWARE
    deviceInfoMap["deviceOSVersion"] = Build.VERSION.RELEASE
    deviceInfoMap["deviceSDKVersion"] = Build.VERSION.SDK_INT
    deviceInfoMap["deviceID"] = Build.ID
    deviceInfoMap["deviceType"] = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
        if (context.checkSelfPermission(Manifest.permission.READ_PHONE_STATE) == PackageManager.PERMISSION_GRANTED) {
            "Phone"
        } else {
            "Phone"
        }
    } else {
        "Unknown"
    }
    deviceInfoMap["localizedModel"] = Build.MODEL 
    deviceInfoMap["deviceLanguage"] = Locale.getDefault().language // Added
    deviceInfoMap["deviceCountry"] = Locale.getDefault().country // Added
    deviceInfoMap["deviceTimeZone"] = TimeZone.getDefault().id // Added
    deviceInfoMap["systemVersion"] = Build.VERSION.RELEASE // Updated
    deviceInfoMap["deviceIMEI"] = Settings.Secure.getString(context.contentResolver, Settings.Secure.ANDROID_ID) // Updated
    deviceInfoMap["secondIMEI"] = getSecondIMEI(context) ?: "Unavailable"
    deviceInfoMap["serialNumber"] = getDeviceSerialNumber()
    deviceInfoMap["macAddress"] = getMacAddress(context) ?: "Unavailable"
    deviceInfoMap["identifierForVendor"] = Settings.Secure.getString(context.contentResolver, Settings.Secure.ANDROID_ID) // Updated
    return deviceInfoMap
}


private fun getMemoryInfo(): Map<String, Any>? {
    val memoryInfo = ActivityManager.MemoryInfo()
    activityManager.getMemoryInfo(memoryInfo)
    val memoryInfoMap = mutableMapOf<String, Any>()
    memoryInfoMap["totalMemory"] = memoryInfo.totalMem
    memoryInfoMap["availableMemory"] = memoryInfo.availMem
    memoryInfoMap["usedMemory"] = memoryInfo.totalMem - memoryInfo.availMem // Calculate used memory
    return memoryInfoMap
}

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}
